app/models/concerns/user_omniauth_concern.rb

```diff
@@ -36,4 +36,15 @@ def create_by_omniauth(auth)
       end
     end
   end
+
+  # Link user with an omniauth provider.
+  #
+  # @param [OmniAuth::AuthHash] auth The data with omniauth info.
+  # @return [Bool] True if success, otherwise false.
+  def link_with_ominiauth(auth)
```

Comment:
in `new_with_session`, call `user.link_with_omniauth(info)` you will likely have to change the contract for link_with_omniauth not to save the record by default, instead the parent resource should save the change.